src/buffer.rs

```diff
@@ -0,0 +1,113 @@
+use std::borrow::{Borrow, Cow};
+use std::io::Read;
+use std::ops::Deref;
+use std::hash::{Hash, Hasher};
+
+use bytes::{Buf, Bytes};
+use string;
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct StrBuf(string::String<Bytes>);
+
+impl StrBuf {
+    pub(crate) fn from_str(s: &str) -> Self {
+        StrBuf(string::String::from_str(s))
+    }
+
+    pub(crate) fn from_utf8(bytes: ByteBuf) -> Result<Self, ::std::str::Utf8Error> {
+        Ok(StrBuf(string::TryFrom::try_from(bytes.0)?))
+    }
+
+    pub(crate) fn from_utf8_lossy(bytes: ByteBuf) -> Self {
+        match String::from_utf8_lossy(bytes.as_ref()) {
+            Cow::Owned(s) => s.into(),
+            Cow::Borrowed(s) => {
+                // SAFETY: We know that `bytes` only contains valid utf-8,
+                //         since the `from_utf8_lossy` operation returned the
+                //         input verbatim.
+                debug_assert_eq!(s.len(), bytes.len());
+                StrBuf(unsafe { string::String::from_utf8_unchecked(bytes.clone().0) })
+            }
+        }
+    }
+}
+
+impl From<String> for StrBuf {
+    fn from(s: String) -> Self {
+        let ByteBuf(bytes) = ByteBuf::from(s.into_bytes());
+        // SAFETY: We know that `bytes` only contains valid utf-8,
+        //         since the underlying data comes from the input string.
+        StrBuf(unsafe { string::String::from_utf8_unchecked(bytes) })
+    }
+}
+
+impl Hash for StrBuf {
```

Comment:
Why not `#[derive]` it? Is there a benefit to matching the `str`'s hash rather than the `Bytes`'s hash (which `String<Bytes>`'s `Hash` forwards to)?

Reply:
Because of the `impl Borrow<str> for StrBuf`, I need to make sure that the hash implementations behave identically between `str` and `StrBuf`.

Quoting the documentation for the [`Borrow`](https://doc.rust-lang.org/std/borrow/trait.Borrow.html) trait:

> Further, when providing implementations for additional traits, it needs to be considered whether they should behave identical to those of the underlying type as a consequence of acting as a representation of that underlying type. Generic code typically uses `Borrow<T>` when it relies on the identical behavior of these additional trait implementations. These traits will likely appear as additional trait bounds.

Without this, it would be impossible to look up an entry from the `names_map` by `&str`, since the `str` and `StrBuf` would evaluate to different hashes, even if they represent the same sequence of chars.